### DISCUSSION TITLE
Animation is  not playing

### DISCUSSION THREAD

**Discussion by F3RN4NDOJF3RN4ND3Z** [StereoKit/StereoKit#244] [Q&A] [answered]
Hi, I am facing some situation playing animations, I load the model and it seems to be getting the animations from the model but when I play it it is not rendering the animation.
Here is what I have
```C#
Model syrenge = Model.FromFile("syrenge.glb");
Pose syrengePose = new Pose(0f, -0.22f, -0.4f, Quat.FromAngles(0,90,0));
syrenge.PlayAnim("pistonAction.002", AnimMode.Loop);
while (SK.Step(() =>{
          UI.Handle("Syrenge", ref syrengePose, syrenge.Bounds );
          syrenge.Draw(syrengePose.ToMatrix());
 })) ;
```

**Comment by maluoi** [comment, marked as answer]
It does look to me like you're doing things right via code! Do you see any warnings in the logs related to the animations? And do you see the animation playing properly in other GLTF viewers [like this one](https://sandbox.babylonjs.com/)?

If that GLTF viewer doesn't play your animation, then it may be an exporter issue on your modeling tool's side. But if it does work, then there could be an issue with StereoKit's animation implementation. If so, I'd love it if I could get a copy of your model so I can check out what might be happening!

**Reply by F3RN4NDOJF3RN4ND3Z** [reply]
I put my model in that  gltf viewer and it plays the animation  properly. I don't see any error message or warnings in the logs, also I added this log
```C#
Log.Info("anim completion:" + syrenge.AnimCompletion);
```
and it prints the current time of the animation as if it were playing I atach the model.
[syrenge.zip](https://github.com/maluoi/StereoKit/files/7636938/syrenge.zip)

**Reply by maluoi** [reply]
Thanks for the file! I can verify, it does play correctly in external viewers, but not in StereoKit. I can see some _very small_ motion in SK, so there may be some scale related issue here. My initial guess is that the scale of some node is not being properly accounted for, but I will investigate further.

If you're interested in an immediate fix while I hunt down the bug, it's possible that restoring the node scale values to 1 may bypass the issue!

**Reply by maluoi** [reply]
I did just find the issue! This is apparently the very first gltf I've found with cubic animation frames, and there's a bug in the cubic animation data parsing/interpolation code :) Now I know what to fix!

**Reply by maluoi** [reply]
Aaand, this is now fixed in the `/dev` branch! I have one more bug I need to fix and then I'll be uploading a new preview build. You can expect that build either Thursday or Friday, but I'll post here when it's up :)

**Reply by maluoi** [reply]
Preview build v0.3.5-preview.3 is now up on NuGet and contains fixes for this! Please try it out and let me know if it fixes things for you :)

**Reply by F3RN4NDOJF3RN4ND3Z** [reply]
Nice ... it works perfectly fine, another question I was trying to achive some animation for a material I am not so good for shaders I used to use shader graph in unity how can I animate a material to achieve an effect like filling that syrenge?

**Reply by maluoi** [reply]
StereoKit does support custom shaders via HLSL, but you'll have to code these effects by hand! The template project has the floor.hlsl shader effect as a reference for how this is added to a project, and you can find additional shader code reference from [the built-in shaders](https://github.com/maluoi/StereoKit/tree/master/StereoKitC/shaders_builtin).

**Comment by F3RN4NDOJF3RN4ND3Z** [comment]
Hi I found that the step is not working or I am not sure if I am ussing it well. I set the Animation Mode to Manual and step if a contition is met but the animation not plays 

in the init method I have
```C#
            syrengeFill =syrenge.FindAnim("pistonAction.002");
            syrenge.PlayAnim(syrengeFill, AnimMode.Manual);
```
And in the Step method I have
```C#
            if (inVial)
            {
                Log.Info("animating:"+syrenge.ActiveAnim.Name + " elapsed:" + syrenge.AnimCompletion);
                if (syrenge.AnimCompletion < 1){
                    syrenge.StepAnim();
                }
                
            }
```

but nothing happens then I modiffy previous code to this
```C#
            if (inVial)
            {
                Log.Info("animating:"+syrenge.ActiveAnim.Name + " elapsed:" + syrenge.AnimCompletion);
                if (syrenge.AnimCompletion == 0){
                    syrenge.PlayAnim(syrengeFill, AnimMode.Once);
                }
                
            }
```
And the animation plays but theres something wonky when the animation reaches the end the piston is disappearing

**Reply by maluoi** [reply]
`StepAnim` should be unnecessary here, its primary purpose is to update the skeleton without, or before calling `Draw` for the `Model`. Instead when using `AnimMode.Manual`, you would set your `syrenge.AnimCompletion` or `syrenge.AnimTime` to the value that you desire each frame.

The piston disappearing at the end of the animation _could_ be a bug. I'll look into it, but I don't think I'm likely to have a fix for that before that hack is over!
